### PR TITLE
Add assignment 1

### PR DESCRIPTION
**Pull request by alexandraryzhevich**

### CODE REVIEW

File: Assignment1Q1.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            return false;
+        }
+        if (str1.isEmpty() && str2.isEmpty()) { // simple case
+            return true;                        // empty strings are permutation of each other
+        }
+
+        Function<String, Map<Character, Integer>> fillCounter = (String str) -> {
+            str = str.toLowerCase();
+            // create hash map to calculate the frequency of each character
+            Map<Character, Integer> counter = new HashMap<>(str.length() * 2);
```
Comment (nfortescue) [inline Assignment1Q1.java:36]: It's nice you have tried to estimate the initial capacity, but it probably isn't worth it. It's a micro-optimisation, the HashMap will still resize in amortized O(1) time. If you are going to estimate the size, then it is a good idea to add a comment explaining your working, including mentioning the load factor.

File: Assignment1Q1.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+     */
+    private static boolean isPermutation(String str1, String str2) {
+        if (!validate(str1, str2)) {
+            return false;
+        }
+        if (str1.isEmpty() && str2.isEmpty()) { // simple case
+            return true;                        // empty strings are permutation of each other
+        }
+
+        Function<String, Map<Character, Integer>> fillCounter = (String str) -> {
```
Comment (nfortescue) [inline Assignment1Q1.java:33]: while Java8 functions are nice, I would have just added this as a static method. Having it as a Function complicates things and doesn't gain you anything that I can see.

File: Assignment1Q1.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                counter.put(c, counter.getOrDefault(c, 0) + 1);
+            }
+            return counter;
+        };
+
+        return fillCounter.apply(str1).equals(fillCounter.apply(str2));
+    }
+
+    /**
+     * More efficient solution for English words.
```
Comment (nfortescue) [inline Assignment1Q1.java:48]: Really good that you are thinking about localization / internationalization here, and commenting this only works for English.

File: Assignment1Q1.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        if (str1.isEmpty() && str2.isEmpty()) { // simple case
+            return true;
+        }
+        str1 = str1.toLowerCase();
+        str2 = str2.toLowerCase();
+
+        // we can use array instead of hash map
+        // because we definitely know amount of characters
+        int[] counter = new int[CHARS_AMOUNT];
+        int startCode = Character.codePointAt("a", 0);
```
Comment (nfortescue) [inline Assignment1Q1.java:63]: If you really understand this, this is really excellent code. But if you wrote this in an interview you should be prepared to answer questions about the difference between codepointAt() and charAt(), and details of Unicode. But knowing the difference is really excellent.

File: Assignment1Q1.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            int checkValue = --counter[idx];
+            if (checkValue < 0) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+
+    public static void main(String[] args) {
```
Comment (nfortescue) [inline Assignment1Q1.java:86]: Really nice you added test code. If you haven't ever seen junit, you might want to try with that.
http://junit.org/junit4/

File: Assignment1Q1.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        System.out.println(isPermutation(null, ""));              // false
+        System.out.println(isPermutation("", ""));                // true
+        System.out.println(isPermutation("a", "a"));              // true
+        System.out.println(isPermutation("a", "ab"));             // false
+        System.out.println(isPermutation("a", "A"));              // true
+        System.out.println(isPermutation("abc", "ACB"));          // true
+        System.out.println(isPermutation("abc", "ABB"));          // false
+        System.out.println(isPermutation("abc", "abd"));          // false
+        System.out.println(isPermutation("ab.", "ab/"));          // false
+        System.out.println(isPermutation("...55,,", ".5.5,.,"));  // true
```
Comment (nfortescue) [inline Assignment1Q1.java:108]: Great set of tests. Others you could have considered:
- A string containing 'Z' / 'z', as it is the last value of the array (boundary condition)
- Strings containing '@', '[', '`' or '(' as they are the boundary conditions for 'a', 'A', 'z' and 'Z' in ASCII
- compare an empty string to a non-empty string

File: Assignment1Q2.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            for (int i = 0; i < k; i++) {
+                if (lead.getNext() == null) {
+                    throw new IndexOutOfBoundsException();
+                }
+                lead = lead.getNext();
+            }
+
+            // now the kth node will be reached
+            // when the lead node will reach the tail
+            SimpleListNode kth = head;
```
Comment (nfortescue) [inline Assignment1Q2.java:54]: Again, a really nice space efficient solution. Well done!

**Review by nfortescue — APPROVED** [review]
